### ISSUE TITLE
addUTR fail on own data post successful run with test3_addUTR.sh

### ISSUE THREAD

**Issue by TGeneralovic** [Gaius-Augustus/BRAKER#297]
Dear BRAKER team/ Katarina and Tom,

Once again thanks for the great toolkit.

I have recently completed (successfully) an annotation using ProtHint and RNAseq evidence followed by UTR prediction using the addUTR argument. 

I have then moved to a new species to repeat this process with new evidence (Proteins for ProtHint and RNAseq) where the run in etp mode completes successfully giving me an expected result. I then attempted to addUTR's in the same manner which failed. No changes where made to program versions or additional conflicting paths installed (gmkey still valid also). 

I am attempting to understand where this failed so I can quickly add the UTR's to this decent annotation. The braker.log file provided the following std output:
```
# Tue Nov 17 17:33:57 2020: rm /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/braker/addUTR/hintsfile.temp.gff
# Tue Nov 17 17:33:57 2020: Adding UTRs to augustus.hints.gtf...
# Tue Nov 17 17:33:57 2020: Running GUSHR...

/usr/bin/python3 /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py -b /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.CAM01.rna.bam /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.SERR01.rna.bam -t /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/braker/augustus.hints.gtf -g /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/braker/addUTR/genome.fa -o /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/braker/addUTR/gushr -c 8 -s /home/tomgeneralovic/BRAKER_dependancies/samtools -a /home/tomgeneralovic/BRAKER_dependancies/augustus-3.3.3/scripts -j /usr/bin -q 2 > /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/braker/addUTR/gushr.log 2> /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/braker/addUTR/errors/gushr.err
```
Where the stated .err file is empty, infact all printed err files seem to indicate no error:
```
`-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 07:51 bam2hints.0.stderr
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 07:51 bam2hints.0.stdout
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 07:51 bam2hints.1.stderr
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 07:51 bam2hints.1.stdout
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 07:51 bam2hints.2.stderr
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 07:51 bam2hints.2.stdout
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 17:33 filterIntronsFindStrand.stderr
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 07:50 find_python3_biopython.err
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 07:50 find_python3_re.err
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 17:33 gushr.err
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 0 Nov 17 17:33 join_mult_hints.rnaseq.stderr
```
So I checked the working directory I told BRAKER to make and work in:

```
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic      4310 Nov 17 07:51 bam_header.map
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic     14490 Nov 17 17:33 braker.log
drwxrwxr-x 2 tomgeneralovic tomgeneralovic      4096 Nov 17 17:33 errors/
drwxrwxr-x 2 tomgeneralovic tomgeneralovic      4096 Nov 17 07:50 GeneMark-ET/
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic 390503579 Nov 17 07:51 genome.fa
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic      4310 Nov 17 07:51 genome_header.map
drwxrwxr-x 2 tomgeneralovic tomgeneralovic    126976 Nov 17 22:33 gushr-GHPJECQCAPHG/
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic      2744 Nov 17 22:33 gushr.log
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic  77568784 Nov 17 17:33 hintsfile.gff
drwxrwxr-x 2 tomgeneralovic tomgeneralovic      4096 Nov 17 07:50 species/
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic         0 Nov 17 17:33 tmp_no_merge_hints.gff
-rw-rw-r-- 1 tomgeneralovic tomgeneralovic      1673 Nov 17 17:33 what-to-cite.txt
```
where it appears that gushr was working away fine until it hit an error, the gushr.log prints:

```
`Searching for samtools:
Will use /home/tomgeneralovic/BRAKER_dependancies/samtools/samtools
Searching for gtf2gff.pl:
Will use /home/tomgeneralovic/BRAKER_dependancies/augustus-3.3.3/scripts/gtf2gff.pl
Trying to execute the following command:
/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools sort -@ 8 /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam -o ./gushr-GHPJECQCAPHG/rnaseq_0_s.bam
Suceeded in executing command.
Error in file /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py at line 281: Return code of subprocess was 1['/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools', 'sort', '-@', '8', '/home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam', '-o', './gushr-GHPJECQCAPHG/rnaseq_0_s.bam']
Will try again...
Trying to execute the following command:
/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools sort -@ 8 /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam -o ./gushr-GHPJECQCAPHG/rnaseq_0_s.bam
Suceeded in executing command.
Error in file /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py at line 281: Return code of subprocess was 1['/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools', 'sort', '-@', '8', '/home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam', '-o', './gushr-GHPJECQCAPHG/rnaseq_0_s.bam']
Will try again...
Trying to execute the following command:
/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools sort -@ 8 /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam -o ./gushr-GHPJECQCAPHG/rnaseq_0_s.bam
Suceeded in executing command.
Error in file /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py at line 281: Return code of subprocess was 1['/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools', 'sort', '-@', '8', '/home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam', '-o', './gushr-GHPJECQCAPHG/rnaseq_0_s.bam']
Will try again...
Trying to execute the following command:
/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools sort -@ 8 /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam -o ./gushr-GHPJECQCAPHG/rnaseq_0_s.bam
Suceeded in executing command.
Error in file /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py at line 281: Return code of subprocess was 1['/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools', 'sort', '-@', '8', '/home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam', '-o', './gushr-GHPJECQCAPHG/rnaseq_0_s.bam']`
```

This indicated that whilst it was struggling to execute samtools it was able to find it in a path eventually but was failing at line 281 of the gushr.py script. To make sure this was not an issue with the version of gushr and samtools I have (even after a sucsessful run previously on another species) I went back and executed the test3.sh followed by the test3_addUTR.sh script which completed without error. Here on in I am thinking there is somthing that gushr is encountering that it is not expecting (in the evidence) and therefore I am unsure on how to fix this. 

THe script used to execute BRAKER was:

```
time    braker.pl \
        --AUGUSTUS_SCRIPTS_PATH=/home/tomgeneralovic/BRAKER_dependancies/augustus-3.3.3/scripts \
        --AUGUSTUS_CONFIG_PATH=/home/tomgeneralovic/BRAKER_dependancies/augustus-3.3.3/config \
        --AUGUSTUS_BIN_PATH=/home/tomgeneralovic/BRAKER_dependancies/augustus-3.3.3/bin \
        --PYTHON3_PATH=/usr/bin/python3.5 \
        --GENEMARK_PATH=/home/tomgeneralovic/BRAKER_dependancies/gmes_linux_64/gmes_petap.pl \
        --BAMTOOLS_PATH=/usr/bin/bamtools \
        --CDBTOOLS_PATH=/home/tomgeneralovic/BRAKER_dependancies/cdbfasta \
        --species=iHelEra_annotation_v2 \
        --genome $REF \
        --addUTR=on \
        --softmasking \
        --bam $BAM1,$BAM2,$BAM3 \
        --workingdir=$WD \
        --AUGUSTUS_hints_preds=augustus.hints.gtf \
        --skipAllTraining \
        --cores $CORES

# Information on time finished
printf "\nDate Finished: $(date)\n\n"
```

I also ensured the  java version installed and activly loaded was compatable with the gushr prog. 
```
gushr.py 1.0.0
AUGUSTUS (3.3.3)
braker.pl version 2.1.5
GeneMark-ES Suite version 4.57_lic
```
Any advice would be greatly apreciated.
Tom Generalovic

**Comment by KatharinaHoff** [comment]
Have you tried with using only 1 core? I implemented that gushr retries
samtools because it often fails at this step in multi core mode. It think
it’s an i/o issue... (the samtools failures) Probably your data is a lot
larger than the test data.

**Comment by TGeneralovic** [comment]
Thanks Katharina, I am repeating with a single core but will take some time. I will update the post when I get some results.

**Comment by TGeneralovic** [comment]
Running with a single core failed with the same error:
```
Searching for samtools:
Will use /home/tomgeneralovic/BRAKER_dependancies/samtools/samtools
Searching for gtf2gff.pl:
Will use /home/tomgeneralovic/BRAKER_dependancies/augustus-3.3.3/scripts/gtf2gff.pl
Trying to execute the following command:
/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools sort -@ 1 /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam -o ./gushr-LHPBIVNCSEAO/rnaseq_0_s.bam
Suceeded in executing command.
Error in file /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py at line 281: Return code of subprocess was 1['/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools', 'sort', '-@', '1', '/home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam', '-o', './gushr-LHPBIVNCSEAO/rnaseq_0_s.bam']
Will try again...
Trying to execute the following command:
/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools sort -@ 1 /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam -o ./gushr-LHPBIVNCSEAO/rnaseq_0_s.bam
Suceeded in executing command.
Error in file /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py at line 281: Return code of subprocess was 1['/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools', 'sort', '-@', '1', '/home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam', '-o', './gushr-LHPBIVNCSEAO/rnaseq_0_s.bam']
Will try again...
Trying to execute the following command:
/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools sort -@ 1 /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam -o ./gushr-LHPBIVNCSEAO/rnaseq_0_s.bam
Suceeded in executing command.
Error in file /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py at line 281: Return code of subprocess was 1['/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools', 'sort', '-@', '1', '/home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam', '-o', './gushr-LHPBIVNCSEAO/rnaseq_0_s.bam']
Will try again...
Trying to execute the following command:
/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools sort -@ 1 /home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam -o ./gushr-LHPBIVNCSEAO/rnaseq_0_s.bam
Suceeded in executing command.
Error in file /home/tomgeneralovic/BRAKER_dependancies/GUSHR/gushr.py at line 281: Return code of subprocess was 1['/home/tomgeneralovic/BRAKER_dependancies/samtools/samtools', 'sort', '-@', '1', '/home/tomgeneralovic/Genome_annotation_BRAKER2/2020-11-07_HelEra_R1/rna_evidence/Hera.AM01.rna.bam', '-o', `'./gushr-LHPBIVNCSEAO/rnaseq_0_s.bam']
```
Maybe I could try breaking the addUTR process down and try each step manually?

**Comment by TGeneralovic** [comment]
As an update I also re-tried with reduced RNA evidence and received the same error.

**Comment by KatharinaHoff** [comment]
I believe this is a samtools issue, not a BRAKER issue. Please refer to the samtools developers for help. I am closing this issue, here. Feel free to re-open.

**Comment by krabapple** [comment]
I am having the same issue with GUSHR.   TGeneralovic , were you able to resolve it?

**Comment by DevonJBoland** [comment]
> I am having the same issue with GUSHR. TGeneralovic , were you able to resolve it?

krabapple, I was having the same issue, and found I had a couple of things that needed to be done to resolve it.

1) I am running on a HPC cluster, and the GUSHR directory was read-only from being installed by an admin. Copying this directory from install location to my personal use directory and changing permission to be read+write, solved the issue with writing out a new GeMoMa.ini.xml file.

2) After that, I was still running into an issue with gushr.py failing and issuing that in line 281 the max number attempts had been reached (default is 4) and it exited with no information in the gushr.stderr file in the errors/ directory. I thought it was an issue with the GeMoMa jar file because the gushr.log had said it successfully completed the samtools steps ( I was using 48 threads). However when I saw this thread and dropped the total number of threads/cores to 8, and ran the modified braker.pl command (training off, etc. in the user manual) after approx. 1 hour, gushr had added the UTR annotations, and created a new augustus.gtf file with the UTR annotations successfully completing the job.

So in short, 1) make sure you have read/write privileges to the GUSHR directory. 2) Samtools is weird, so lower your --cores value to 10% of total cpus (I had 80 core nodes hence 8 cores) and try re-running it. Hope this helps, I have been spending all week trying to get UTR annotations, as this is one of the best programs to do so (IMO).
